overlord/snapshotstate/snapshotmgr.go

```diff
@@ -339,6 +338,8 @@ func doRestore(task *state.Task, tomb *tomb.Tomb) error {
 	restoreState.Config = oldCfg
 	task.Set("restore-state", restoreState)
 
+	backendCleanup(restoreState)
```

Comment:
this is all confusing, but I suspect we can't simply do this because it breaks undos?

Reply:
as undoRestore will want to use restoreState. I don't know if we have relevant tests or not for this :/

Reply:
> the problem is that as designed a restore change can have many restore tasks, some might succeed but if one fails everything is undone, but if we do this, we lost the state of the one that succeeded when we try to undo all of them

But does this really matter here? I guess that a restore change can have many restore tasks *on different snaps*, right? In our specific case, the cleanup handler goal is to get rid of the temporary files created during the restore operation, so I don't think it matters whether it's executed when all tasks are done, or right after the corresponding task is done: the end result will be the same.

I might need to add a call to `backendCleanup()` also on the undo task (I've yet to verify if it's needed), but other than that I don't see an issue with this. Or can you make an example of when things could go wrong?

Reply:
the issue is that in many snaps restore where one fails,  with this code we end up calling Revert on restoreState that has been Cleanup for the snaps that succeeded, and that can/will fail